packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+        } else if (tableSize === 0) {
+          validationError(EOFError.RJUMPVTableSize0)
+        }
+
+        if (ptr + tableSize * 2 + 2 >= code.length) {
+          // Fall-through case
+          validationError(EOFError.OpcodeIntermediatesOOB)
+        }
+
+        const newPc = ptr + 2 + tableSize * 2
+
+        for (let i = 0; i < tableSize; i++) {
+          const newPtr = ptr + 2 + i * 2
+          // Add the table bytes to intermediates
+          addIntermediate(newPtr)
+          addIntermediate(newPtr + 1)
+          const target = readInt16(code, newPtr) + newPc
+          if (target < 0 || target >= code.length) {
+            validationError(EOFError.OpcodeIntermediatesOOB)
+          }
+
+          successorSet.add(target)
+
+          addJump(target)
+          reachableOpcodes.add(target)
+        }
+
+        // Special case for RJUMPV: move ptr over the table (the immediate starting byte will be added later)
+        // In this special case, add the immediate starting byte
+        addIntermediate(ptr + 1)
+        ptr += 2 * tableSize + 1
+      } else if (opcode === 0xe3 || opcode === 0xe5) {
+        // CALLF / JUMPF
+        const target = readUint16(code, ptr + 1)
+        reachableSections[codeSection].add(target)
+        if (target >= container.header.codeSizes.length) {
+          validationError(EOFError.InvalidCallTarget)
+        }
+        if (opcode === 0xe3) {
+          // CALLF
+          const targetOutputs = container.body.typeSections[target].outputs
+          const targetInputs = container.body.typeSections[target].inputs
+          if (targetOutputs === 0x80) {
+            // CALLF points to non-returning function which is not allowed
+            validationError(EOFError.InvalidCALLFReturning)
+          }
+
+          if (minStackCurrent < targetInputs) {
+            validationError(EOFError.StackUnderflow)
+          }
+
+          if (
+            maxStackCurrent + container.body.typeSections[target].maxStackHeight - targetInputs >
+            1024
+          ) {
+            validationError(EOFError.StackOverflow)
+          }
+
+          minStackNext += targetOutputs - targetInputs
+          maxStackNext += targetOutputs - targetInputs
+        } else {
+          // JUMPF
+          const currentOutputs = container.body.typeSections[codeSection].outputs
+          const targetOutputs = container.body.typeSections[target].outputs
+          const targetInputs = container.body.typeSections[target].inputs
+          const targetNonReturning = targetOutputs === 0x80
+
+          if (targetOutputs > currentOutputs && !targetNonReturning) {
+            // Spec rule:
+            // JUMPF operand must point to a code section with equal or fewer number of outputs as
+            // the section in which it resides, or to a section with 0x80 as outputs (non-returning)
+            validationError(EOFError.InvalidJUMPF)
+          }
+
+          if (returningFunction && targetOutputs <= 0x7f) {
```

Comment:
Is this invalid?  As I've noted before, I'm pretty sure that `returningFunction` is actually "false" if the function returns anything so not sure if this check is correct or not.

Reply:
Sorry the terminology is confusing. We are currently in a code section which has 0x80 outputs (so it is "non returning" by the EOF spec). However, the target of the JUMPF is now returning. This means that if you CALLF a 0x80 code section, and this JUMPFs into this <= 0x7F (returning code section), then thus CALLF did not call into a non-returning function, but it actually called into a function, and therefore this invalidates the container.